// @kimonolabs/models/models/publicapi.js
'use strict';

var mongoose = require('mongoose');
var Schema = mongoose.Schema;

module.exports = function(db) {
  var SourceApiSchema = new Schema({
    name:String,
    source:String,
    id:String,
    idx:Number,
    apiref: {type: Schema.Types.ObjectId, ref: 'Api' }
  });

  var CollectionSchema = new Schema({
    _id:String,
    name:String,
    desc:String,
    type:String, //{api|source|user}
    idx:Number,
    propertydesc:Object
  });

  // === schema ===
  var PublicApiSchema = new Schema({
    name: String,
    id: String,
    desc: String,
    apis: [SourceApiSchema],
    collections: [CollectionSchema],
    active: Boolean,
    callcount: Number,
    createdon: Date,
    createdby: String,
    userref: {type: Schema.Types.ObjectId, ref: 'User' }
  });

  // mongoose.model('SourceApi', SourceApiSchema);
  // mongoose.model('Property', PropertySchema);
  // mongoose.model('Collection', CollectionSchema);
  db.model('PublicApi', PublicApiSchema);
  return db.model('PublicApi');
};
